File: alliander_robotics/alliander_robotiq/src/alliander_robotiq/src/finger_status.cpp
// SPDX-FileCopyrightText: Alliander N. V.
//
// SPDX-License-Identifier: Apache-2.0

#include "alliander_robotiq/finger_status.hpp"

#include <array>
#include <sstream>
#include <stdexcept>

#include "alliander_robotiq/json_utils.hpp"

int FingerStatus::number_of_joints() const { return scissor ? 2 : 3; }

void FingerStatus::update_status(std::uint8_t status_index) {
  static const std::array<std::string, 4> statuses = {
      "in_motion", "contact_while_opening", "contact_while_closing",
      "reached_target_position"};

  if (status_index >= statuses.size()) {
    throw std::out_of_range("Invalid finger status index.");
  }

  status = statuses[status_index];
}

std::string FingerStatus::to_json() const {
  std::ostringstream out;
  out << "{";
  out << "\"scissor\":" << (scissor ? "true" : "false") << ",";
  out << "\"status\":" << json_string(status) << ",";
  out << "\"position_request\":" << position_request << ",";
  out << "\"position\":" << position << ",";
  out << "\"current\":" << current;
  out << "}";
  return out.str();
}

File: alliander_robotics/alliander_robotiq/src/alliander_robotiq/alliander_robotiq/finger_status.py
# SPDX-FileCopyrightText: Alliander N. V.
#
# SPDX-License-Identifier: Apache-2.0

import typing
from dataclasses import dataclass
from typing import Literal, Optional

from mashumaro.mixins.json import DataClassJSONMixin

STATUS_FINGER = Literal[
    "in_motion",
    "contact_while_opening",
    "contact_while_closing",
    "reached_target_position",
]


@dataclass
class FingerStatus(DataClassJSONMixin):
    """Dataclass to hold the status of a specific finger.

    Attributes:
        scissor (bool): Whether the finger is a scissor or not.
        status (Optional[STATUS_FINGER]): The status of the finger.
        position_request (Optional[int]): The requested position of the finger.
        position (Optional[int]): The position of the finger.
        current (Optional[int]): The current of the finger.
    """

    scissor: bool = False
    status: Optional[STATUS_FINGER] = None
    position_request: Optional[int] = None
    position: Optional[int] = None
    current: Optional[int] = None

    def update_status(self, status: int) -> None:
        """Update the status of the finger.

        Args:
            status (int): The status as an integer.
        """
        self.status = typing.get_args(STATUS_FINGER)[status]

    @property
    def number_of_joints(self) -> int:
        """The number of joints for the finger.

        Returns:
            int: The number of joints (2 for scissor, 3 for regular fingers).
        """
        return 2 if self.scissor else 3
